dali/pipeline/data/sample_view.h

```diff
@@ -0,0 +1,157 @@
+
+/**
+ * @brief SampleView - non-owning view type, keeping the pointer, shape and runtime type information
+ * of specific sample.
+ *
+ * This is a type corresponding to Tensor<Backend> and it's main goal is to become the return type
+ * for the TensorVector/TensorList::operator[]. It allows to access the data via the
+ * convenient `view<T, ndim>(SampleView)` conversion to TensorView, but doesn't break the batch
+ * object encapsulation and doesn't allow to adjust the allocation.
+ */
+template <typename Backend>
+class SampleView {
+ public:
+  /**
+   * @name Get the underlying pointer to data
+   */
+  // @{
+  /**
+   * @brief Return an un-typed pointer to the underlying storage.
+   * A valid type must be set prior to calling this method by calling
+   * the non-const version of the method, or calling 'set_type'.
+   */
+  void *raw_mutable_data() {
+    return data_;
+  }
+
+  /**
+   * @brief Return a const, un-typed pointer to the underlying storage.
+   * A valid type must be set prior to calling this method by calling
+   * the non-const version of the method, or calling 'set_type'.
+   */
+  const void *raw_data() const {
+    return data_;
+  }
+
+  /**
+   * @brief Returns a typed pointer to the underlying storage.
+   * The calling type must match the underlying type of the buffer.
+   */
+  template <typename T>
+  inline T *mutable_data() {
+    DALI_ENFORCE(
+        type() == TypeTable::GetTypeId<T>(),
+        make_string(
+            "Calling type does not match buffer data type, requested type: ",
+            TypeTable::GetTypeId<T>(), " current buffer type: ", type(),
+            ". To set type for the Buffer use 'set_type<T>()' or Resize(shape, type) first."));
+    return static_cast<T *>(data_);
+  }
+
+  /**
+   * @brief Returns a const, typed pointer to the underlying storage.
+   * The calling type must match the underlying type of the buffer.
+   */
+  template <typename T>
+  inline const T *data() const {
+    DALI_ENFORCE(
+        type() == TypeTable::GetTypeId<T>(),
+        make_string(
+            "Calling type does not match buffer data type, requested type: ",
+            TypeTable::GetTypeId<T>(), " current buffer type: ", type(),
+            ". To set type for the Buffer use 'set_type<T>()' or Resize(shape, type) first."));
+    return static_cast<T *>(data_);
+  }
+  //@}
+
+  /**
+   * @brief Get the shape of the sample
+   */
+  const TensorShape<> &shape() const {
+    return shape_;
+  }
+
+  /**
+   * @brief Get the runtime information about the type of the sample
+   */
+  DALIDataType type() const {
+    return type_id_;
+  }
+
+
+  SampleView() = default;
+
+  SampleView(const SampleView &) = default;
+  SampleView &operator=(const SampleView &) = default;
+
+  SampleView(SampleView &&other) {
+    data_ = other.data_;
+    other.data_ = nullptr;
+    shape_ = std::move(other.shape_);
+    other.shape_ = {};
+    type_id_ = other.type_id_;
+    other.type_id_ = DALI_NO_TYPE;
+  }
```

Comment:
Reuse implementation?
```suggestion
  SampleView(SampleView &&other) {
    *this = std::move(other);
  }
```